db_setup/upload_seed_data.py

```diff
@@ -24,87 +26,63 @@ def get_database_engine():
     """Seeds database with plant data."""
+    plant_table = db.Table('plant', metadata, autoload_with=engine)
+
     try:
-        conn.execute(sql.text(f"USE {environ['DB_NAME']};"))
-        conn.execute(sql.text("BEGIN TRANSACTION;"))
-
-        for row in plants:
-            if row["origin_location"]:
-                row["origin_location"] = int(float(row["origin_location"]))
-            query = sql.text(
-                f"""INSERT INTO {environ['DB_SCHEMA']}.plant (name, scientific_name, location_id)
-                    VALUES (:name, :scientific_name, :origin_location);""")
-            conn.execute(query, row)
-
-        conn.execute(sql.text("COMMIT;"))
+        conn.execute(plant_table.insert(), plants)
         conn.commit()
-
     except Exception as e:
-        conn.execute(sql.text("ROLLBACK;"))
+        conn.rollback()
         raise e
 
 
-def upload_duties(conn: Connection, duties: list) -> None:
+def upload_duties(duties: list, engine: db.Engine, conn: Connection, metadata) -> None:
     """Seeds database with information about each botanists responsibility."""
-    try:
-        conn.execute(sql.text(f"USE {environ['DB_NAME']};"))
-        conn.execute(sql.text("BEGIN TRANSACTION;"))
-
-        for row in duties:
-            query = sql.text(
-                f"""INSERT INTO {environ['DB_SCHEMA']}.duty (plant_id, botanist_id)
-                    VALUES (:plant_id, :botanist_id);""")
-            conn.execute(query, row)
+    duty_table = db.Table('duty', metadata, autoload_with=engine)
 
-        conn.execute(sql.text("COMMIT;"))
+    try:
+        conn.execute(duty_table.insert(), duties)
         conn.commit()
-
     except Exception as e:
-        conn.execute(sql.text("ROLLBACK;"))
+        conn.rollback()
         raise e
 
 
 def upload() -> None:
     """Combines each function to seed the database."""
-    with open('seed_locations.csv', 'r', encoding="utf-8") as csv_file:
-        locations = list(csv.DictReader(csv_file))
 
-    with open('seed_plants.csv', 'r', encoding="utf-8") as csv_file:
-        plants = list(csv.DictReader(csv_file))
+    locations = pd.read_csv('seed_locations.csv').replace(np.nan, None).to_dict('records')
+
+    plants_df = pd.read_csv('seed_plants.csv')
+    plants_df['origin_location'] = plants_df['origin_location'].fillna(-1).astype(int)
+    plants_df = plants_df.replace({np.nan: None, -1: None})
+    plants = plants_df.to_dict('records')
 
-    with open('seed_duties.csv', 'r', encoding="utf-8") as csv_file:
-        duties = list(csv.DictReader(csv_file))
+    duties = pd.read_csv('seed_duties.csv').replace(np.nan, None).to_dict('records')
 
     engine = get_database_engine()
     conn = engine.connect()
+    metadata = db.MetaData(schema=environ['DB_SCHEMA'])
 
-    upload_locations(conn, locations)
+    upload_locations(locations, engine, conn, metadata)
 
-    upload_plants(conn, plants)
+    upload_plants(plants, engine, conn, metadata)
```

Comment:
you know more than me with this but do you actually have to pass in engine and connection, could you not just pass in the engine?

Reply:
We need it to commit / rollback inside the function – we could always do that this in this function, or declare a new connection from the engine inside each function, but the first offers worse compartmentalisation of transactions, and the second would be a lot slower, and a bit redundant (I believe).

Reply:
okay cool makes sense